### ISSUE TITLE
Support for method argument names in nameof()

### ISSUE THREAD

**Issue by taori** [dotnet/csharplang#373]
![image](https://cloud.githubusercontent.com/assets/5545184/24409469/9531f634-13d0-11e7-8f18-51b42f5b2506.png)

it would be fairly amazing if nameof also worked for MethodName arguments.

I think i've talked about this 1-2 years ago and some aspnet member also chimed in and mentioned how this could be useful for magic EntityFramework where the appropriate name must be passed too for connection strings to work.

So essentially i would love it if, instead of strings i could be typing this:

```
	[AuthGrantFilter(AppFunction = ApplicationFunctionTypes.ReadProjectStructure, TopLevelProjectId = nameof(topLevelProjectId), SiteId = nameof(siteId))]
		[AuthDenyFilter(Context = DenyContext.Site, TopLevelProjectId = nameof(topLevelProjectId), SiteId = nameof(siteId))]
```
or this

```
	[AuthGrantFilter(AppFunction = ApplicationFunctionTypes.ReadProjectStructure, TopLevelProjectId = GetChildrenOf.nameof(topLevelProjectId), SiteId = GetChildrenOf.nameof(siteId))]
		[AuthDenyFilter(Context = DenyContext.Site, TopLevelProjectId = GetChildrenOf.nameof(topLevelProjectId), SiteId = GetChildrenOf.nameof(siteId))]
```

**Comment by jnm2** [comment]
Why is `GetChildrenOf.nameof(topLevelProjectId)` better than `nameof(GetChildrenOf.topLevelProjectId)`? Putting `nameof` first makes it easier to read.

**Comment by jnm2** [comment]
Or `nameof(GetChildrenOf, topLevelProjectId)`.

**Comment by alrz** [comment]
How would you resolve multiple method overloads? It's not ambiguous in the use-site, but when you rename either of parameters in the method declaration, it's not clear which one `nameof` is referring to.

**Comment by jnm2** [comment]
@alrz `nameof(GetChildrenOf(int, int?), topLevelProjectId)` perhaps.

**Comment by alrz** [comment]
That seem a lot of new syntax, however, the attributes in the example above is on the method itself. I think it would be nice to just bring them into the scope, e.g.
```cs
[Attribute(nameof(parameter))]
void M(object parameter, [Attribute(nameof(parameter))] object p) { }
```

**Comment by jnm2** [comment]
I agree.

**Comment by taori** [comment]
@jnm2 you could a typo of me there unfortunately. i meant to write it like nameof(GetChildrenOf.siteId). I'd rather stay in line with the ordinary nameof syntax

**Comment by taori** [comment]
@alrz I would probably resolve it like this given we have the following methods

```
class Sample {
Method1(string a, string b) {}
Method2(string a, string b) {}
Method2(string a, string b, string c) {}
Method2(int aint, int bint, int cint) {}
}
```

Which would resolve in options to use nameof like this:

```
// methodname unique, not need to specify signature to remain uniquely identifyable
nameof(Sample.Method1.a) nameof(Sample.Method1.b)

nameof(Sample.Method2.a, string, string) nameof(Sample.Method2.b, string, string)

nameof(Sample.Method2.a, string, string, string) nameof(Sample.Method2.b, string, string, string) nameof(Sample.Method2.c, string, string, string)

// argument names unique - no need to be more specific to stay unique
// given that i don't quite like this one from an intellisense perspective since "Method2" should then suggest "aint" and "a" at the same time
nameof(Sample.Method2.aint) nameof(Sample.Method2.bint) nameof(Sample.Method2.cint)
```

**Comment by taori** [comment]
> This would be great for unit testing methods throwing ArgumentExceptions.

Quoting @perholje https://github.com/dotnet/csharplang/issues/771#issuecomment-320895137

**Comment by cartermp** [comment]
cc @jcouv 

Note that the `NotNullIfNotNull(string)`, when specified for the return value, makes this a bit more desirable:

```csharp
class Path
{
    [return: NotNullIfNotNull(nameof(path))] // This fails to compile, needs to be "path" instead
    public static string? GetFileName(string? path);
}
```

In fact, the LDM notes inadvertently specified code that won't compile due to this 🙂 - https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-05-15.md#nullness-dependencies-between-inputs-and-outputs

**Comment by CyrusNajmabadi** [comment]
The main challenge here is back compat.  How do you do this without breaking existing code.  You'd need a way to say "try with the existing binding first, and if that doesn't work, look at parameters".  But then that's *super* weird and will not match any sort of intuition about what should be going on when i see something like that.

**Comment by TylerBrinkley** [comment]
Just putting my 2 cents in as well that this would be useful, especially with the addition of the `NotNullIfNotNullAttribute` as @cartermp points out. It wouldn't be surprising to me if more attributes like that get added in future versions for the nullable reference types feature.

**Comment by RikkiGibson** [comment]
@jaredpar brought up a scenario could break depending on how we implement this:

```cs
class A : System.Attribute
{
    public A(int i, string s) { }
}

class C
{
    const int item1 = 123;
    int item2;

    [A(item1, nameof(item2))]
    void M(int item1, int item2) { }
}
```

If we don't specify/implement it carefully, item1 will start binding to the parameter and the code will no longer compile. So one suggestion is that we could attempt to bind the attribute argument using basically the "binder we already have today", and if the lookup fails, attempt to bind again using a binder that knows about the method parameters. Apologies if I'm mangling the terms here.

Alternatively we could decide no, the user needs to change their code to say `[A(C.item1, nameof(item2))]` instead, specifying the constant member of `C` instead of the parameter of `M`.

**Comment by Joe4evr** [comment]
> If we don't specify/implement it carefully, item1 will start binding to the parameter and the code will no longer compile.

Is it possible to specify it'll only bind to a parameter of the method an attribute is applied on if it's inside a `nameof()` expression?

**Comment by taori** [comment]
@RikkiGibson That sure is an interesting edge case he came up with. In that scenario i would however say that the feature should just raise a compile error if it can't distinguish between symbols. After all you are free to pick and choose argument names without major impacts on your code, so utilizing the benefits of forwarding argument names to attributes would still exist even if you have that case.

+ afaik naming convention for most private members is still a leading underscore, so there should be no collisions if you follow conventions.

**Comment by yaakov-h** [comment]
@taori that's just one naming convention, and not even the default VS one.

**Comment by taori** [comment]
@yaakov-h Yeah - sure. Well i guess the language team will wrap their head around this one since there seems to be quite some interest in implementing this feature to ensure compilation safety for the nullness safety feature.

**Comment by CyrusNajmabadi** [comment]
> Is it possible to specify it'll only bind to a parameter of the method an attribute is applied on if it's inside a nameof() expression?

Yes.  That would be my preference.

**Comment by Joe4evr** [comment]
> Allow `nameof` to reference parameters with the method name as a qualifier `[Attr(nameof(M.p))] void M(int p) { }`

Funny thing, I thought of *this exact option* to be the final disambiguator, if nothing else will do. I even had thought out that the user experience when in the context `[Attr(nameof(M.$$))]` should be to have IntelliSense pop up regarding `M` as a "quasi-object" (hopefully simple to do in Roslyn's design) and listing M's parameters as its members.

**Comment by gafter** [comment]
We probably would want to do this either before or at the same time as https://github.com/dotnet/csharplang/issues/287

**Comment by AartBluestoke** [comment]
option 3 could have a problem of concerning class members. It is likely we could have a class of the same name as a function in scope (eg: decorating a constructor). 
This is more likely than having a class member of the same name as a function parameter name in scope as almost all code styles would name members and locals differently.

```
Class C{
    int my_int;
    // do you expect my_int, the_int, both or neither to work here?
    [Attr(nameof(C.my_int),nameof(C.the_int))] 
    C(int the_int){ ... };
}
```

**Comment by jcouv** [comment]
We discussed this in LDM this week and landed on option [2] (special scoping rules for nameof in attributes). This will be backwards compatible.

**Comment by mikernet** [comment]
This would be rather useful for some NRT annotations:
```c#
[return: NotNullIfNotNull(nameof(source))]
public static string? Resolve(object? source)
```

**Comment by batzen** [comment]
How about `nameof(parameters.xy)` or something like that?
That would prevent all scoping issues.

**Comment by 333fred** [comment]
@batzen what if I have a field named `parameters` with a property or field named `xy`?

**Comment by batzen** [comment]
Sorry, meant `params`...

**Comment by CommonLoon102** [comment]
With the new `CallerArgumentExpression` this is needed even more than before:
https://github.com/dotnet/roslyn/discussions/57792

**Comment by jcouv** [comment]
Feature was merged for VS 17.3p2
